Add several fields to advocacy campaigns and outreaches

irskep opened the pull request:
# Summary

Advocacy campaign:
* templates
* postal_addresses
* email_addresses
* phone_numbers

Outreach:
* notes
* postal_addresses
* email_addresses
* phone_numbers

Also fixes a typo.

# Motivation

## Advocacy campaigns

### Templates

It's very common for grassroots organizations to send weekly emails with call scripts for members of Congress. These scripts can be captured in the description field, but by pulling them out into a separate field (here generalized to "templates"), you give clients an opportunity to highlight them.

Templates can be used for advocacy call scripts, or writing templates for postcards, emails, or letters.

### Postal addresses, email addresses, phone numbers

When organizations send out weekly actions, they very often include phone numbers inline. From a data entry or scraper standpoint, forcing all of these into Person objects will either create a huge number of duplicate people, force all the data into the description field, or just lose it altogether.

Also, sometimes people are contacting an organization or an office more than a specific person.

This can also function as a schema safety valve for all the cases we haven't thought of for contact info related to advocacy campaigns.

## Outreaches

### Notes

People sometimes report back if something interesting happened on an advocacy campaign call. This is just a basic text field to capture that.

### Addresses, phone numbers

This is meant to capture which option an outreach-er chose from the options available in the campaign. Using this data, it might be possible for people to determine which offices are most available week to week, and which ones are not effective, so people publishing campaigns can prioritize the contact info they send out.

# Implementation

I just updated the Markdown docs. I didn't see anything else that needed to happen.

I didn't try to build them. Wanted to get feedback first.

irskep commented:
I think this covers everything we discussed on the call and within RagTag. Happy to do more!

j-ro commented:
This looks pretty good, thanks! A few thoughts:

- If there are only three types of contact methods (phone, address, email), maybe we want templates to be an object that has a template for each type? template.phone, template.email, template.address? Otherwise, I'm unsure why it's an array...

- Maybe outreaches should have templates too? It's fairly common to have an advocacy campaign that targets multiple targets (ie, all of congress) with customized scripts for each outreach/specific target. So maybe outreaches have templates too, or maybe the template lives within the target object?

- Same thing for the added fields on outreaches, especially the contact method -- should that live in the target object, since it's the method used on the target?

irskep commented:
> If there are only three types of contact methods (phone, address, email), maybe we want templates to be an object that has a template for each type? template.phone, template.email, template.address? Otherwise, I'm unsure why it's an array...

Members of Congress have multiple offices. For example, Diane Feinstein has SF, LA, and DC offices. Actions need phone numbers for all three because the mailboxes fill up.

> Maybe outreaches should have templates too? It's fairly common to have an advocacy campaign that targets multiple targets (ie, all of congress) with customized scripts for each outreach/specific target. So maybe outreaches have templates too, or maybe the template lives within the target object?

The target is just a string right now and I was wary of adding a new type. I figured if people wanted to change the script per target, they'd just create more than one campaign.

> Same thing for the added fields on outreaches, especially the contact method -- should that live in the target object, since it's the method used on the target?

This makes sense if we create a target object instead of just using a string.

> Finally, we'll want to add the sub-object tables for postal addresses, emails, phones, etc.. as we do on our standard objects other places, but that's just a small editing note.

I don't quite understand what you mean, but I assume it's some docs I forgot to add, which I will correct when I figure it out. :-)

j-ro commented:
Targets are already objects I believe:

http://opensupporter.github.io/osdi-docs/outreaches.html#target

irskep commented:
Oh, I missed that! Cool! On advocacy campaigns it's a string. I'll move it over and update the targets object definition on both.

j-ro commented:
Yeah, on advocacy campaigns it's supposed to be "A human readable description of the target universe for this advocacy campaign. (ex: “U.S. Congress”)", so could just be the name there and that's fine, but on outreaches you get more. Maybe we want to expand target on advocacy campaigns into an object too though.

j-ro commented:
you able to join us on the tech call now to discuss? email me at jason@actionnetwork.org if you need the link to lucid.

irskep commented:
An outreach just represents one person doing one thing to one target, right? So if I were to display a list of campaigns with the numbers you could call, where would those go?

j-ro commented:
Right, so those should maybe go in advocacy campaigns, so maybe we want to expand what we do with targets there...

irskep commented:
I moved the contact info into the Target object and added a list of targets to campaigns. According to my notes, this is the extent of the changes we discussed on the call. Any other improvements I can make?

j-ro commented:
Looks good to me! A couple editing things (like maybe put the email addresses/phones/physical addresses objects in your subobject include?), but seems like what we discussed!

irskep commented:
I also just realized I didn't update any of the example JSON responses. Seems important.

joshco commented:
Yes, the examples are important, but they are quasi editorial.  We can review and approve the meat of the proposal as long as you can follow on with the updated examples for next week.

irskep commented:
Sounds good.

I'm a little unclear on what @j-ro means by put the "email addresses/phones/physical addresses objects in your subobject include" - aren't they there already?

irskep commented:
Just after we got off the call, I realized I couldn't find the setup/build instructions. Is it just 'install jekyll and run `jekyll` in the root'?

joshco commented:
check https://jekyllrb.com/

irskep commented:
Got it building, markup seems to be OK, I'll put it online hopefully later this evening.

irskep commented:
* Moved phone/email/postal address to its own include (from `people.md`)
* Added it to `target_subobject` so it appears in `outreaches.md` and `advocacy_campaigns.md`

irskep commented:
Changes are visible on my gh-pages branch, which lives here:

* http://steveasleep.com/osdi-docs/advocacy_campaigns.html
* http://steveasleep.com/osdi-docs/outreaches.html
* http://steveasleep.com/osdi-docs/people.html

j-ro commented:
Just editing stuff, so can be approved without, but:

- postal addresses et al should be linked to their respective anchors here: http://steveasleep.com/osdi-docs/outreaches.html#target (see how it's done on people)
- same with the same sections on advocacy campaigns
- should probably edit some of the examples to show this expanded format

j-ro commented:
one more:

- target universe seems to be gone from advocacy campaigns, but it stays, right? we should have both, since one can be used if it's like 500 folks like congress, and the other if it's one or a handful. So we should have both, and the language in the definition should disambiguate between them?

irskep commented:
I didn't make any changes related to target universe. Do I need to rebase?

j-ro commented:
Hm, not sure, we can look into that after merge then -- you can see it in the examples though.

irskep commented:
Added examples. But it looks like my git config is a little broken and GitHub doesn't know how to handle it, so I'm going to squash my commits and get a cleaner history. How important is commit message quality to you?

irskep commented:
Fixed; that should be everything except target universe. I did the examples pretty quickly, let me know if they need to be improved. They should have the right schema though!

j-ro commented:
Wanna rerun your jekyll so we can see over in your live examples?

irskep commented:
Just pushed changes to gh-pages, should be live in a minute.

j-ro commented:
Nice. Some small spacing issues in the example here: http://steveasleep.com/osdi-docs/outreaches.html

And I'd maybe not do inline comments, since we largely don't otherwise. I'd just make your points clear in the description in the table maybe, and I think they largely already are.

Need to figure out about that missing universe field, but otherwise, looking good!

irskep commented:
Can you be more specific about the spacing issues? I'm not sure what you're talking about.

j-ro commented:
Look at postal_addresses in the target in the first example -- some space vs tab issues perhaps in the JSON

joshco commented:
__Accidental breaking change__

We need to pick a new attribute name for the new target list array to avoid a breaking change.

Old Version:
 |targets	|string	|A human readable description of the target universe for this advocacy campaign. (ex: "U.S. Congress") 

New Version:
 |targets			|[Target[]](#target)    |An array of target object hashes representing the targets of the outreach.

Perhaps the new attribute should be "target_list"?

irskep commented:
Oh, great catch!!

And thanks for pointing out the spacing issue, I wasn't even looking at the examples.

joshco commented:
ETA?

hypeJunction commented:
@j-ro Is your branch the definitive next iteration of the spec? We are a few weeks away from launching OSDI compliant API for civ.works (which is tightly integrated with activismengine.org), so I'd rather work of the next iteration of the spec if it's 'ready'.

j-ro commented:
This is being merged soon, yes. It has already been approved, though this branch still needs some editing work for consistency and such. If you'd like more details, I'd suggest you join the coalition and our tech call weekly so we can discuss in real time: http://opensupporter.org/

j-ro commented:
We try, yeah -- we have folks joining us from India as well. They're at 11am Eastern on Thursdays.

hypeJunction commented:
I am interested in OSDI only for the duration of the project (really find hateoas hard to work with and luckily we are building this from scratch, otherwise integrating it into an existing API would be a problem, though mapping the existing data in and out is already a pain). I might join you for the next few meetings.

j-ro commented:
Great! We got your application here, our folks should follow up soon.

hypeJunction commented:
Here is a use case that won't fit into an existing model (let me know if I should open a new issue). We allow advocacy campaigns that dynamically populate representative information by looking it up in Google's Civic Information API. So the country-wide campaign encourages to reach out to Senate or House Rep / Commission / School Board etc on a Federal / State / County / City level without enumerating the details of individual offices/persons to contact. Depending on user's location, the contact details are populated during action rendering.

j-ro commented:
That's what the target universe field is for, so you can say what the overall universe is you're targeting. Then, on each outreach, you can say which target was contacted based on location information.

hypeJunction commented:
You may want to update your branch. You have targets listed twice once as string, once as Target[]. I suppose the first needs to be target_universe.

j-ro commented:
Yes, that's why this has not been merged yet and is a work in progress.

hypeJunction commented:
I made a PR to rename original target to target universe. My personal opinion is that the field is too broad and multiple providers will not be able to reconcile, i.e. `Supreme Court` is too ambiguous and could mean a `Supreme Course of the United States of America`, `Supreme Course of the People's Republic of China` or `Supreme Court of Alabama` or any other of the 50 states. Turning it into a hash with the location field and ocdid would be a gain.

hypeJunction commented:
Also I think a `Target` is really an extension of a `Person`, so perhaps it should be treated as such. You want to have all the honorific prefixes etc with the names, as well as other profiles. Perhaps using `osdi:targets` in `_embedded` would be a more of a HAL way to do this.

j-ro commented:
Please don't go for a new PR here, we are working on the other one. Targets and people are not semantically similar, though they have similar fields.

hypeJunction commented:
In an SQL system, I am finding it hard to keep targets in sync, e.g. if the target is removed from the list, there is no way of reliably knowing which one it was. I would encourage moving target to hal embeds and assigning them identifiers.

j-ro commented:
Right now, I'd say this would be best as a namespaced, proprietary extension. This has come up a few times on the committee call and hasn't moved forward because there aren't multiple members interested in implementing it (usually a prerequisite for getting in the spec) and it presents some complicated technical challenges in working with the API, the most important of which POST operations become multi-step POST and GETs, where you have to look up IDs, POST them to advocacy campaigns or outreaches, and in generally make everything a lot more complicated and chatty.

So, for now, I'd say if you want to push forward with IDs and links, you should do it namespaced. I'm sure folks here would be happy to look at your proposal and help make it OSDI-fluent so it could be easily merged into the spec at a later date if the approach became more popular.

hypeJunction commented:
For now, I am comparing records by given and last name, and if they don't match, I wipe up, otherwise update them. The only downside, is we are loosing any interactions with that record if the name changes or some reason. Also a problem with targets that don't have names.
Namespacing doesn't make much sense, because we will be mostly consuming APIs, so it's more about mapping external data to our records.

j-ro commented:
Well, given that no other OSDI member is interested in implementing targets with IDs, I think you'll have to do that matching for now.

hypeJunction commented:
Sorry for engaging you on issues that you may have discussed and decided upon, but just throwing things out there, so they are in the back of your heads.
We have advocacy campaigns that request people to tweet to a specific person or post on their Facebook page. It seems adding `Profiles[]` as in the person item would be helpful for `Target`

j-ro commented:
Sure, yes, though that would be a separate proposal.

irskep commented:
Hi, sorry for dropping off the map for a bit, I'm working on this now.

irskep commented:
It would help me if the off-topic discussion could happen in another place so I don't keep getting notifications for things not directly relevant to this patch.

irskep commented on outreaches.md:
```diff
@@ -238,14 +269,14 @@ Cache-Control: max-age=0, private, must-revalidate
+                        "organization": "U.S. Senate",
+                        "ocdid": "ocd-division/country:us/state:va"
+                    }
+                },
```
These examples all had a single tab at the beginning.

irskep commented:
* Rebased on current `gh-pages`
* Fixed inconsistent indentation
* Renamed new `targets` to `target_list` _only on advocacy campaigns_, since outreaches already have them

I'll re-read the thread to make sure I haven't missed anything!

joshco commented:
Approved cmte 5/25
